internal/directives/http_requester.go

```diff
@@ -0,0 +1,253 @@
+			fmt.Errorf("error building HTTP request: %w", err)
+	}
+	resp, err := h.getClient(cfg).Do(req)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+	env, err := h.buildExprEnv(resp)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error building expression context from HTTP response: %w", err)
+	}
+	success, err := h.wasRequestSuccessful(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating success criteria: %w", err)
+	}
+	failure, err := h.didRequestFail(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating failure criteria: %w", err)
+	}
+	switch {
+	case success && !failure:
+		outputs, err := h.buildOutputs(cfg.Outputs, env)
+		if err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("error extracting outputs from HTTP response: %w", err)
+		}
+		return PromotionStepResult{
+			Status: kargoapi.PromotionPhaseSucceeded,
+			Output: outputs,
+		}, nil
+	case failure:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseFailed},
+			fmt.Errorf("HTTP response met failure criteria")
+	default:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseRunning}, nil
+	}
+}
+
+func (h *httpRequester) buildRequest(cfg HTTPConfig) (*http.Request, error) {
+	method := cfg.Method
+	if method == "" {
+		method = http.MethodGet
+	}
+	req, err := http.NewRequest(method, cfg.URL, bytes.NewBufferString(cfg.Body))
+	if err != nil {
+		return nil, fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	for _, header := range cfg.Headers {
+		req.Header.Add(header.Name, header.Value)
+	}
+	if len(cfg.QueryParams) > 0 {
+		q := req.URL.Query()
+		for _, queryParam := range cfg.QueryParams {
+			q.Add(queryParam.Name, queryParam.Value)
+		}
+		req.URL.RawQuery = q.Encode()
+	}
+	return req, nil
+}
+
+func (h *httpRequester) getClient(cfg HTTPConfig) *http.Client {
+	httpTransport := cleanhttp.DefaultTransport()
+	if cfg.InsecureSkipTLSVerify {
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true, // nolint: gosec
+		}
+	}
+	return &http.Client{
+		Transport: httpTransport,
+	}
+}
+
+func (h *httpRequester) buildExprEnv(resp *http.Response) (map[string]any, error) {
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %w", err)
+	}
+	env := map[string]any{
+		"response": map[string]any{
+			"status":  resp.StatusCode,
+			"header":  resp.Header.Get,
+			"headers": resp.Header,
+			"body":    map[string]any{},
+		},
+	}
+	if len(bodyBytes) > 0 && resp.Header.Get(contentTypeHeader) == contentTypeJSON {
```

Comment:
We should probably account for the content type header containing a charset (i.e. `application/json; charset=utf-8`).

Reply:
Fixed in 01077dbe3d50dc9e52426fc48f43a3123b03e56e